sol/main.c

```diff
@@ -435,63 +430,98 @@ req_queue_init(struct sol_config *sol_conf)
 	 * the integer number of cycles per byte to the numerator.
 	 */
 	cycles_per_byte_precise = cycles_per_sec / max_credit_bytes_precise;
-	req_queue->cycles_per_byte_floor = cycles_per_byte_precise;
+	cycles_per_byte_floor = cycles_per_byte_precise;
 	ret = rte_approx(
-		cycles_per_byte_precise - req_queue->cycles_per_byte_floor,
+		cycles_per_byte_precise - cycles_per_byte_floor,
 		sol_conf->tb_rate_approx_err, &a, &b);
 	if (ret < 0) {
 		SOL_LOG(ERR, "Could not approximate the request queue's allocated bandwidth\n");
 		return ret;
 	}
-	req_queue->cycles_per_byte_a = a;
-	req_queue->cycles_per_byte_b = b;
 
-	/* Add integer number of cycles per byte to numerator. */
-	req_queue->cycles_per_byte_a +=
-		req_queue->cycles_per_byte_floor * req_queue->cycles_per_byte_b;
+	for (i = 0; i < sol_conf->num_lcores; i++) {
+		struct req_queue *req_queue = &sol_conf->instances[i].req_queue;
+
+		req_queue->len = 0;
+		req_queue->highest_priority = 0;
+		req_queue->lowest_priority = GK_MAX_REQ_PRIORITY;
+
+		/* Initialize token bucket as full. */
+		req_queue->tb_max_credit_bytes = round(max_credit_bytes_precise);
+		req_queue->tb_credit_bytes = req_queue->tb_max_credit_bytes;
+		req_queue->cycles_per_byte_floor = cycles_per_byte_precise;
+		req_queue->cycles_per_byte_a = a;
+		req_queue->cycles_per_byte_b = b;
+
+		/* Add integer number of cycles per byte to numerator. */
+		req_queue->cycles_per_byte_a +=
+			req_queue->cycles_per_byte_floor * req_queue->cycles_per_byte_b;
 
-	SOL_LOG(NOTICE, "Cycles per byte (%f) represented as a rational: %"PRIu64" / %"PRIu64"\n",
-		cycles_per_byte_precise,
-		req_queue->cycles_per_byte_a, req_queue->cycles_per_byte_b);
+		SOL_LOG(NOTICE, "Cycles per byte (%f) represented as a rational: %"PRIu64" / %"PRIu64" at lcore %u\n",
+			cycles_per_byte_precise,
+			req_queue->cycles_per_byte_a, req_queue->cycles_per_byte_b,
+ 			sol_conf->lcores[i]);
+
+		req_queue->time_cpu_cycles = rte_rdtsc();
+	}
 
-	req_queue->time_cpu_cycles = rte_rdtsc();
 	return 0;
 }
 
 static int
 cleanup_sol(struct sol_config *sol_conf)
 {
-	struct req_queue *req_queue = &sol_conf->req_queue;
-	struct rte_mbuf *entry, *next;
+	int i;
 
-	list_for_each_entry_safe_m(entry, next, &req_queue->head) {
-		list_del(mbuf_to_list(entry));
-		rte_pktmbuf_free(entry);
-		req_queue->len--;
-	}
+	for (i = 0; i < sol_conf->num_lcores; i++) {
+		struct req_queue *req_queue = &sol_conf->instances[i].req_queue;
+		struct rte_mbuf *entry, *next;
```

Comment:
Notice that `cleanup_sol()` may be called from `sol_stage1()`, so field `sol_conf->instances` may be `NULL`. This condition would trigger a segmentation fault here.